formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt

```diff
@@ -0,0 +1,300 @@
+@ExperimentalSerializationApi
+public sealed class CborElement(
+    /**
+     * CBOR tags associated with this element.
+     * Tags are optional semantic tagging of other major types (major type 6).
+     * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+     */
+    @OptIn(ExperimentalUnsignedTypes::class)
+    tags: ULongArray = ulongArrayOf()
+
+) {
+    /**
+     * CBOR tags associated with this element.
+     * Tags are optional semantic tagging of other major types (major type 6).
+     * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+     */
+    @OptIn(ExperimentalUnsignedTypes::class)
+    public var tags: ULongArray = tags
+        internal set //need this to collect
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is CborElement) return false
+
+        if (!tags.contentEquals(other.tags)) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        return tags.contentHashCode()
+    }
+
+}
+
+/**
+ * Class representing CBOR primitive value.
+ * CBOR primitives include numbers, strings, booleans, byte arrays and special null value [CborNull].
+ */
+@Serializable(with = CborPrimitiveSerializer::class)
+@ExperimentalSerializationApi
+public sealed class CborPrimitive<T : Any>(
+    public val value: T,
+    tags: ULongArray = ulongArrayOf()
+) : CborElement(tags) {
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is CborPrimitive<*>) return false
+        if (!super.equals(other)) return false
+
+        if (value != other.value) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = super.hashCode()
+        result = 31 * result + value.hashCode()
+        return result
+    }
+
+    override fun toString(): String {
+        return "${this::class.simpleName}(" +
+            "tags=${tags.joinToString(prefix = "[", postfix = "]")}, " +
+            "value=$value" +
+            ")"
+    }
+}
+
+/**
+ * Class representing either:
+ * * signed CBOR integer (major type 1 encompassing `-2^64..-1`)
+ * * unsigned CBOR integer (major type 0 encompassing `0..2^64-1`)
+ *
+ * depending on the value of [sign]. Note that [absoluteValue] **must** be `0` when sign is set to [Sign.ZERO]
+ */
+@Serializable(with = CborIntSerializer::class)
+@ExperimentalSerializationApi
+public class CborInt(
+    absoluteValue: ULong,
+    public val sign: Sign,
+    vararg tags: ULong
+) : CborPrimitive<ULong>(absoluteValue, tags) {
+
+    init {
+        if (sign == Sign.ZERO) require(absoluteValue == 0uL) { "Illegal absolute value $absoluteValue for Sign.ZERO" }
+    }
+
+    public enum class Sign {
+        POSITIVE,
+        NEGATIVE,
+        ZERO
+    }
+
+    /**
+     * **WARNING! Possible truncation/overflow!** E.g., `-2^64` -> `1`
```

Comment:
Silent truncation/overflow is something we should avoid. I'd vote for an exception if a value could not be represented as `Long` (`Int`, `Short`, or `Byte` - more on that later), and an additional `toLongOrNull` function, that'll return `null` on overflow.

It would be nice to align the API with what we already have for `JsonElement` and instead of providing a member function, provide extension properties.
And it seems reasonable to provide properties returning other integer types too.